lib/logger/lib/logger.ex

```diff
@@ -463,41 +441,39 @@ defmodule Logger do
   """
   @spec metadata(metadata) :: :ok
   def metadata(keyword) do
-    {enabled?, metadata} = __metadata__()
-    Process.put(@metadata, {enabled?, into_metadata(keyword, metadata)})
-    :ok
-  end
+    case :logger.get_process_metadata() do
+      :undefined ->
+        :ok = :logger.set_process_metadata(filter_out_nils(keyword))
 
-  defp into_metadata([], metadata), do: metadata
-  defp into_metadata(keyword, metadata), do: into_metadata(keyword, [], metadata)
+      map when is_map(map) ->
+        merged = Enum.into(keyword, map)
+        metadata = filter_out_nils(merged)
```

Comment:
We can probably do the two operations above (merge and filter out) in one pass. So please either do it in one pass OR leave a TODO so we tackle it later. :)